### PR TITLE
Improvement: Generate bindings in parallel

### PR DESCRIPTION
**Pull request by ferozco**
## Before this PR
We would sequentially generate an emit each file for a given conjure definition. This was particularly painful for some very larger internal projects with thousands of objects where it could take up to 24s to emit (on my local machine)

## After this PR

==COMMIT_MSG==
Generate bindings in parallel
==COMMIT_MSG==

After parallelization we've seen some big perf wins

### CODE REVIEW

File: conjure-java/src/main/java/com/palantir/conjure/java/cli/ConjureJavaCli.java [hunk 1]
```diff
@@ -187,26 +190,24 @@ public void run() {
...
-                    ServiceGenerator undertowGenerator = new UndertowServiceGenerator(config.options());
-                    undertowGenerator.emit(conjureDefinition, config.outputDirectory());
+                    generatorBuilder.add(new UndertowServiceGenerator(config.options()));
                 }
                 if (config.generateDialogue()) {
-                    ServiceGenerator dialogueServiceGenerator = new DialogueServiceGenerator(config.options());
-                    dialogueServiceGenerator.emit(conjureDefinition, config.outputDirectory());
+                    generatorBuilder.add(new DialogueServiceGenerator(config.options()));
                 }
+                new GenerationCoordinator(Executors.newCachedThreadPool(), generatorBuilder.build())
```
Comment (carterkozak) [inline conjure-java/src/main/java/com/palantir/conjure/java/cli/ConjureJavaCli.java]: daemon threads + shutdown in a finally

File: conjure-java-core/src/main/java/com/palantir/conjure/java/GenerationCoordinator.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * Generates and emits to the given output directory all services and types of the given conjure definition, using
+     * the instance's service and type generators.
+     */
+    public List<Path> emit(ConjureDefinition conjureDefinition, File outputDir) {
+        return MoreStreams.inCompletionOrder(
+                        generators.stream().flatMap(generator -> generator.generate(conjureDefinition)),
+                        f -> Goethe.formatAndEmit(f, outputDir.toPath()),
+                        executor,
+                        32)
```
Comment (carterkozak) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/GenerationCoordinator.java]: Should we use something dynamic based on the core count?
```suggestion
                        Runtime.getRuntime().availableProcessors())
```

Comment (ferozco) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/GenerationCoordinator.java]: sure!

File: conjure-java-core/src/main/java/com/palantir/conjure/java/GenerationCoordinator.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+     * Generates and emits to the given output directory all services and types of the given conjure definition, using
+     * the instance's service and type generators.
+     */
+    public List<Path> emit(ConjureDefinition conjureDefinition, File outputDir) {
+        return MoreStreams.inCompletionOrder(
+                        generators.stream().flatMap(generator -> generator.generate(conjureDefinition)),
+                        f -> Goethe.formatAndEmit(f, outputDir.toPath()),
+                        executor,
+                        Runtime.getRuntime().availableProcessors())
+                .collect(Collectors.toList());
```
Comment (carterkozak) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/GenerationCoordinator.java:49]: Do we need to unwrap thrown exceptions, or does the MoreStreams utility handle that for us? (sorry, I'm not very familiar with the library!)

Comment (ferozco) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/GenerationCoordinator.java:49]: its handled for us :)

**Review by carterkozak — APPROVED** [review]
thanks!

**Pull request merged**